### PR TITLE
TipBox.vue: Support different size icons 

### PR DESCRIPTION
**Pull request by yash-chowdhary**
closes [MarkBind #986](https://github.com/MarkBind/markbind/issues/986) (specifications of sizes can also be found there). 

This is an enhancement to box icons.
The following classes can be made use of when adding icons to boxes - `icon-medium`, `icon-large`, `icon-extra-large`.

### CODE REVIEW

**Comment by yash-chowdhary** [comment]
Ready for Review

File: src/TipBox.vue [hunk 1]
```diff
@@ -185,11 +209,11 @@
     .icon-wrapper {
         display: flex;
         margin-right: .5em;
-        width: 22px;
-        height: 22px;
     }
 
     .contents {
+        display: flex;
+        align-items: center;
```
Comment (acjh) [inline src/TipBox.vue]: Why are these changes necessary?

Comment (yash-chowdhary) [reply src/TipBox.vue]: This was added to vertically center-align the content of the box.
Here's what the content will look like without the vertical alignment -  
Screenshot 2020-01-23 at 9 37 59 PM

Vertically aligned content -
Screenshot 2020-01-23 at 9 41 06 PM

Comment (acjh) [reply src/TipBox.vue]: Did you try with long paragraphs? I don't think these changes are appropriate.

Comment (damithc) [reply src/TipBox.vue]: Test for different length content to make sure the solution work for all situations (in general, it's safer to use realistic data for examples/tests). Use fontawsome icons when giving screenshots.

Comment (yash-chowdhary) [reply src/TipBox.vue]: > Did you try with long paragraphs? I don't think these changes are appropriate.

Yup, you're right. This was causing problems for paragraphs. I've modified the css class for proper vertical center aligment now and tested it out with large paragraphs and single-line content.

Tests with different icon sizes -
1. small icon; single-line 
Screenshot 2020-01-23 at 11 40 01 PM

* medium icon; single-line
Screenshot 2020-01-23 at 11 44 08 PM

* large icon; single-line
Screenshot 2020-01-23 at 11 44 45 PM

* extra-large icon; single-line
Screenshot 2020-01-23 at 11 45 17 PM

* medium icon; multi-line
Screenshot 2020-01-24 at 12 04 26 AM

* large icon; multi-line
Screenshot 2020-01-24 at 12 05 11 AM

* extra-large icon; multi-line
Screenshot 2020-01-24 at 12 05 29 AM

Comment (damithc) [reply src/TipBox.vue]: Looks OK to me.

File: src/TipBox.vue [hunk 2]
```diff
@@ -185,11 +185,14 @@
...
-        width: 22px;
-        height: 22px;
     }
 
     .contents {
+        display: flex;
+        flex-direction: column;
+        justify-content: center;
+        align-content: center;
+        font-size: medium;
```
Comment (acjh) [inline src/TipBox.vue]: I believe we only need `display` and `align-content`.

File: src/TipBox.vue [hunk 3]
```diff
@@ -209,4 +212,16 @@
     .font-black {
         color: #24292e;
     }
+
+    .icon-extra-large {
```
Comment (acjh) [inline src/TipBox.vue]: Specify `.icon-extra-large i` to not affect other elements.

Though as I mentioned in the comments that I linked earlier, such arbitrarily styling should really be in the site's own CSS file, rather than included in MarkBind/vue-strap.

File: src/TipBox.vue (hunk 2 above)
Comment (yash-chowdhary) [reply src/TipBox.vue]: I tried it without the `flex-direction: column`, and the `justify-content: center`. Here's what it looks like for large paragraphs - 
Screenshot 2020-01-24 at 5 16 13 PM

Here's what it looks like without `justify-content: center` (but with `flex-direction: column`). While it does work for long paragraphs, single-line content is not properly aligned -
Screenshot 2020-01-24 at 5 19 51 PM

File: src/TipBox.vue (hunk 3 above)
Comment (yash-chowdhary) [reply src/TipBox.vue]: > Though as I mentioned in the comments that I linked earlier, such arbitrarily styling should really be in the site's own CSS file, rather than included in MarkBind/vue-strap.

Oh, I see. I misunderstood thinking it was only about the usage of "icon-large" instead. Will make the change.

**Comment by yash-chowdhary** [comment]
As mentioned above, usage:
```html
<box type="warning" add-class="icon-large">
    DUMMY TEXT
</box>
```

Will open a PR in markbind to add the CSS icon-size classes to the assets.

**Comment by damithc** [comment]
> As mentioned above, usage:
> 
> ```
> <box type="warning" add-class="icon-large">
>     DUMMY TEXT
> </box>
> ```

Oh? I thought it is going to be `<box type="warning" icon-large>`

**Comment by yash-chowdhary** [comment]
I thought @acjh wanted to normalize it to using `add-class`?

**Comment by damithc** [comment]
> I thought @acjh wanted to normalize it to using `add-class`?

@acjh what do you think? `<box type="warning" icon-large>` is easier to the author but I don't mind using `<box type="warning" add-class="icon-large">`  if there a justification.

**Comment by acjh** [comment]
> Will open a PR in markbind to add the CSS icon-size classes to the assets.

By *the site's own CSS file*, I meant like https://github.com/nus-cs2103/website-base/blob/master/css/main.css.

> @acjh what do you think? `<box type="warning" icon-large>` is easier to the author but I don't mind using `<box type="warning" add-class="icon-large">` if there a justification.

The styles are arbitrary (2em, 3em, extra-large). If we do this, then authors may expect MarkBind to add more styles like medium-large (2.5em), extra-extra-large (5em) and text sizes next. *There is little value that MarkBind can add*, especially since these can be achieved with plain CSS quite easily. *MarkBind should go for pleasant styling*, maybe with alternative defaults if reasonable, but we should draw the line between supporting the ability to do something (add-class and the other changes in this PR) and adding arbitrary styles.

File: src/TipBox.vue (hunk 2 above)
Comment (acjh) [reply src/TipBox.vue]: That appears to be because `<slot></slot>` is directly in `contents`, together with `heading` and `dismiss-button`.

I don't think it is appropriate to always centre the children in `contents` though, particularly the `heading`. Did we try `extra-large` (maybe extremely-large) with `heading` and/or `dismissable`?

**Comment by damithc** [comment]
> > Will open a PR in markbind to add the CSS icon-size classes to the assets.
> 
> By _the site's own CSS file_, I meant like https://github.com/nus-cs2103/website-base/blob/master/css/main.css.
> 
> > @acjh what do you think? `<box type="warning" icon-large>` is easier to the author but I don't mind using `<box type="warning" add-class="icon-large">` if there a justification.
> 
> The styles are arbitrary (2em, 3em, extra-large). If we do this, then authors may expect MarkBind to add more styles like medium-large (2.5em), extra-extra-large (5em) and text sizes next. _There is little value that MarkBind can add_, especially since these can be achieved with plain CSS quite easily. _MarkBind should go for pleasant styling_, maybe with alternative defaults if reasonable, but we should draw the line between supporting the ability to do something (add-class and the other changes in this PR) and adding arbitrary styles.

I agree with drawing the line somewhere. I'm also mindful that our target authors are expected to be able to use basic html-like syntax only whereas css editing is moving to web developer territory. As an author, I myself sometimes ask you guys for css hacks when I can't get MarkBind to do what I want. It's not something we can expect our target users to do (at least not the authors of education materials; might be OK for those writing project documentation if the project itself is a web project). So, it's good if we can minimize the need for css editing. 

For example, the box icon size vary between different site generators (ours seem to be the smallest)  but probably the 4 proposed variations can cover the full range. That way, our box can be configured to be close to most of the common box styles. If the author wants some odd-sized icon, only then they should have to resort to css editing. 

Do you think it is OK to provide some built-in styles, like how Bootstrap does (e.g., `text-danger text-info pr-1 pr-2` etc.)? That way, our users can kind of make their own styles without having to write css code themselves. That is, to have something like this in the UG:

> You can vary the icon size by adding `icon-medium`,  `icon-large`,  `icon-extralarge` to a box via the `add-class` mechanism. If those sizes don't fit your needs, you can set it to your own size as follows:
> Add a class to your site's custom css file.
>Example:
>```css
>.icon-custom i {
>    font-size: 3em;
>}
>```
>Then, ...

**Comment by acjh** [comment]
> Do you think it is OK to provide some built-in styles, like how Bootstrap does (e.g., `text-danger text-info pr-1 pr-2` etc.)?

I think that is not where MarkBind should be heading.

Some investigation shows that Font Awesome already does sizing:
https://fontawesome.com/how-to-use/on-the-web/styling/sizing-icons

`fa-lg` is `1.33333em` — *large* is arbitrary.

How about:

```html
<!-- Raw -->
<box type="warning" icon=":fas-camera:" icon-size="2x">
    DUMMY TEXT
</box>

<!-- Rendered -->
<div class="alert container alert-warning">
  <div class="icon-wrapper fa-2x">
    <span><i class="fas fa-camera"></i></span>
  </div>
  <div class="contents">
    DUMMY TEXT
  </div>
</div>
```

Note that `fa-{iconSize}` is added to `icon-wrapper` rather than the icon element `i` itself, unlike the Font Awesome examples, to simplify support for custom Markdown `icon`.

**Comment by damithc** [comment]
`<box type="warning" icon=":fas-camera:" icon-size="2x">`  sounds good.

File: src/TipBox.vue (hunk 2 above)
Comment (yash-chowdhary) [reply src/TipBox.vue]: Sorry for the late reply. I've been trying out different things to make this work. You're right, the `heading` and `dismiss-button` don't behave well when I center-align the contents. Center-aligning the contents is much harder than I thought it would be :sweat_smile:

I've tried wrapping the `<slot>` in a `<div>` and `<span>`, but that doesn't work well. The next thing I tried was moving the `heading` and `button` outside the 'content's `div` and make it a 3 column container (one for the icon, one for text, one for heading and button). The problem here is that the `heading` messes with the text column if it is too long. 
So, the latest thing I'm working on is separating out the `heading` into a `div` that would be placed on top of the container, effectively making it a 2-rowed container. The 2nd row will be a 3 column container, with the 3rd column only for the `dismiss-button`. 

Are there any better ways to deal with this?

Comment (acjh) [reply src/TipBox.vue]: I would say not to vertically center-align the contents.

Comment (yash-chowdhary) [reply src/TipBox.vue]: Alright 👍

**Comment by yash-chowdhary** [comment]
Usage: 
```html
<box type="warning" icon=":fas-camera:" icon-size="2x">
    DUMMY TEXT
</box>
```

(Possible) Expected output(s):
* icon-size="2x", single-para
  Screenshot 2020-01-30 at 10 03 24 PM


* `icon-size="2x"`, multi-para
  Screenshot 2020-01-30 at 10 02 02 PM

* `icon-size="2x"`, multi-para, no heading, not dismissible
  Screenshot 2020-01-30 at 10 05 48 PM

File: src/TipBox.vue [hunk 4]
```diff
@@ -1,16 +1,18 @@
 <template>
     <div class="alert container" :class="[boxStyle, addClass, lightStyle]" :style="customStyle">
-        <div v-if="!isDefault" class="icon-wrapper">
+        <div v-if="!isDefault && !iconSize" class="icon-wrapper">
+            <span v-html="iconType"></span>
+        </div>
+        <div v-if="!isDefault && iconSize" class="icon-wrapper" :class="`fa-${iconSize}`" >
```
Comment (acjh) [inline src/TipBox.vue]: Let's do as done with `lightStyle`:

```html
<div v-if="!isDefault" class="icon-wrapper" :class="[iconStyle]">
```

File: src/TipBox.vue [hunk 5]
```diff
@@ -1,16 +1,18 @@
...
-            <h6 v-if="heading" class="heading">{{ heading }}</h6>
-            <button v-if="dismissible" type="button" class="close dismiss-button" data-dismiss="alert" aria-label="Close">
-                <span aria-hidden="true">&times;</span>
-            </button>
-            <slot></slot>
+          <h6 v-if="heading" class="heading">{{ heading }}</h6>
+          <button v-if="dismissible" type="button" class="close dismiss-button" data-dismiss="alert" aria-label="Close">
+            <span aria-hidden="true">&times;</span>
+          </button>
+          <slot></slot>
```
Comment (acjh) [inline src/TipBox.vue]: Please revert this unnecessary and inconsistent dedentation.

File: src/TipBox.vue [hunk 6]
```diff
@@ -31,6 +33,10 @@
         type: String,
         default: null
       },
+      iconSize: {
+        type: String,
+        default: null
+      },
```
Comment (acjh) [inline src/TipBox.vue]: If no semantic nor alphabetical order, then default to adding at the end.

In this case, we can put `iconSize` after `icon` ⁠— grouping `icon*` props and then sorting by alphabetical order.

File: src/TipBox.vue [hunk 7]
```diff
@@ -1,6 +1,6 @@
 <template>
     <div class="alert container" :class="[boxStyle, addClass, lightStyle]" :style="customStyle">
-        <div v-if="!isDefault" class="icon-wrapper">
+        <div v-if="!isDefault" class="icon-wrapper" :class="[`fa-${iconSize}`]" >
```
Comment (acjh) [inline src/TipBox.vue]: Wouldn't this result in `class="fa-null"` if `iconSize` not specified?

Comment (yash-chowdhary) [reply src/TipBox.vue]: Fixed

File: src/TipBox.vue [hunk 8]
```diff
@@ -1,6 +1,6 @@
 <template>
     <div class="alert container" :class="[boxStyle, addClass, lightStyle]" :style="customStyle">
-        <div v-if="!isDefault" class="icon-wrapper">
+        <div v-if="!isDefault" class="icon-wrapper" :class="`${iconSize ? `fa-${iconSize}`: ''}`" >
```
Comment (acjh) [inline src/TipBox.vue]: Let's do as done with `lightStyle` as suggested in https://github.com/MarkBind/vue-strap/pull/124#discussion_r372991071:

```html
<div v-if="!isDefault" class="icon-wrapper" :class="[iconStyle]">
```

File: src/TipBox.vue [hunk 9]
```diff
@@ -1,6 +1,6 @@
 <template>
     <div class="alert container" :class="[boxStyle, addClass, lightStyle]" :style="customStyle">
-        <div v-if="!isDefault" class="icon-wrapper">
+        <div v-if="!isDefault" class="icon-wrapper" :class="[iconStyle]" >
```
Comment (acjh) [inline src/TipBox.vue]: Remove stray whitespace:

```suggestion
        <div v-if="!isDefault" class="icon-wrapper" :class="[iconStyle]">
```

File: src/TipBox.vue [hunk 10]
```diff
@@ -146,9 +149,14 @@
...
             return '<i class="fas fa-exclamation"></i>';
         }
+      }, 
+      iconStyle() {
+        if (this.iconSize) {
+          return `fa-${this.iconSize}`;
       }
     }
   }
+  }
```
Comment (acjh) [inline src/TipBox.vue]: - Misplaced closing brace.
- Missing default return value.
```diff
      iconStyle() {
        if (this.iconSize) {
          return `fa-${this.iconSize}`;
+       }
+       return '';
      }
    }
  }
- }
```
